src/algebra/lie/weights.lean

```diff
@@ -231,37 +233,142 @@ end lie_module
       submodule.coe_smul_of_tower, linear_map.coe_comp, linear_map.sub_apply], },
   simp_rw [linear_map.commute_pow_left_of_commute (hfg _)],
 end
 
+variables {H M}
+
+lemma lie_mem_weight_space_of_mem_weight_space {χ₁ χ₂ : H → R} {x : L} {m : M}
+  (hx : x ∈ root_space H χ₁) (hm : m ∈ weight_space M χ₂) : ⁅x, m⁆ ∈ weight_space M (χ₁ + χ₂) :=
+begin
+  apply lie_module.weight_vector_multiplication
+    H L M M ((to_module_hom R L M).restrict_lie H) χ₁ χ₂,
+  simp only [lie_module_hom.coe_to_linear_map, function.comp_app, linear_map.coe_comp,
+    tensor_product.map_incl, linear_map.mem_range],
+  use [⟨x, hx⟩ ⊗ₜ ⟨m, hm⟩],
+  simp only [submodule.subtype_apply, to_module_hom_apply, submodule.coe_mk,
+    lie_module_hom.coe_restrict_lie, tensor_product.map_tmul],
+end
+
+variables (R L H M)
+
+/-- Given a nilpotent Lie subalgebra `H ⊆ L` together with `χ₁ χ₂ : H → R`, there is a natural
+`R`-linear product of root vectors and weight vectors, compatible with the actions of `H`. -/
+def root_space_weight_space_product (χ₁ χ₂ : H → R) :
+  (root_space H χ₁) ⊗[R] (weight_space M χ₂) →ₗ⁅R,H⁆ weight_space M (χ₁ + χ₂) :=
+lift_lie R H (root_space H χ₁) (weight_space M χ₂) (weight_space M (χ₁ + χ₂))
+{ to_fun    := λ x,
+  { to_fun    :=
+      λ m, ⟨⁅(x : L), (m : M)⁆, lie_mem_weight_space_of_mem_weight_space x.property m.property⟩,
+    map_add'  := λ m n, by { simp only [lie_submodule.coe_add, lie_add], refl, },
+    map_smul' := λ t m, by { conv_lhs { congr, rw [lie_submodule.coe_smul, lie_smul], }, refl, }, },
+  map_add'  := λ x y, by ext m; rw [linear_map.add_apply, linear_map.coe_mk, linear_map.coe_mk,
+    linear_map.coe_mk, subtype.coe_mk, lie_submodule.coe_add, lie_submodule.coe_add, add_lie,
+    subtype.coe_mk, subtype.coe_mk],
+  map_smul' := λ t x, by ext m; rw [linear_map.smul_apply, linear_map.coe_mk, linear_map.coe_mk,
+    subtype.coe_mk, lie_submodule.coe_smul, smul_lie, lie_submodule.coe_smul, subtype.coe_mk],
+  map_lie'  := λ x y, by ext m; rw [lie_hom.lie_apply, lie_submodule.coe_sub, linear_map.coe_mk,
+    linear_map.coe_mk, subtype.coe_mk, subtype.coe_mk, lie_submodule.coe_bracket,
+    lie_submodule.coe_bracket, subtype.coe_mk, lie_subalgebra.coe_bracket_of_module,
+    lie_subalgebra.coe_bracket_of_module, lie_submodule.coe_bracket,
+    lie_subalgebra.coe_bracket_of_module, lie_lie], }
+
+@[simp] lemma coe_root_space_weight_space_product_tmul
+  (χ₁ χ₂ : H → R) (x : root_space H χ₁) (m : weight_space M χ₂) :
+  (root_space_weight_space_product R L H M χ₁ χ₂ (x ⊗ₜ m) : M) = ⁅(x : L), (m : M)⁆ :=
+by simp only [root_space_weight_space_product, lift_apply, lie_module_hom.coe_to_linear_map,
+  coe_lift_lie_eq_lift_coe, submodule.coe_mk, linear_map.coe_mk, lie_module_hom.coe_mk]
+
+/-- Given a nilpotent Lie subalgebra `H ⊆ L` together with `χ₁ χ₂ : H → R`, there is a natural
+`R`-linear product of root vectors, compatible with the actions of `H`. -/
+def root_space_product (χ₁ χ₂ : H → R) :
```

Comment:
One thing that might turn out to be really annoying is that you know have `χ₁ + χ₂` in the target, which is not defeq to `χ₂ + χ₁`. Similarly if you chain these products together, you will get associativity issues, etc. I guess this already happens with `root_space_weight_space_product` as well.

Reply:
I guess there are at least two approaches around this:

1. Take a third argument `χ3` and a proof `h : χ₁ + χ₂ = χ3`
2. Only take the `χ3` and just return `0` if `χ₁ + χ₂ = χ3` is false. (This can be justified by saying that `root_space_product χ₁ χ₂ χ3` is the composition of "inject into the graded algebra of root spaces, take the product, and project down to root space `χ3`.)

Reply:
Hmm, very good point. Option 2 is in particular is tempting but at the very least I think I should look more closely at the approach to gradings that @eric-wieser and @kbuzzard have developed as part of LTE.

I've committed (to myself) to spend the today refereeing an article but I'll try to give this some thought in parallel.

Reply:
I have now adjusted this (choosing option 1 in the end).